Bluetooth: Controller: Add CTE reception and sampling in connected mode

ppryga-nordic opened the pull request:
To make possible to correctly receive PDUs that optionally include
Constant Tone Extension after CRC the PHYEND event must be used
instead of END event. PHYEND event is generated by Radio peripheral
for actual end of a PDU. In case there is a CTE the PHYEND event
is generated at end of the CTE. If PDU does not have CTE, then
PHYEND event is generated at end of CRC. In this case it is the
same instant as END event.

Use of PHYEND event is required only when BLE 5.1 Receiving Constant
Tone Extensions feature is supported and BT_CONN is enabled.

Radio peripheral that includes Direction Finding Extension is
able to generate EVENTS_PHYEND at very end of received PDU.
In case there is a Constant Tone Extensions in received PDU
the EVENTS_PHYEND event is generated after end of CTE reception.
If there is not CTE present in received PDU, the EVENTS_PHYEND
event is generate in the same instant as EVENTS_END.

If CTEINLINE is enabled then Radio will do a runtime PDU parsing
to check if CTEInfo is present in received PDU. In case there
is no CTEInfo the EVENTS_PHYEND event will be generated with
16 us delay after EVENTS_END. 

The PR also adds possibility to receive and sample CTE. 
The samples are collected and reported to host.
Sampling is done only if requested by host.

ppryga-nordic committed:
Bluetooth: Controller: Use PHYEND in SW switch for nRF52 SOCs with DFE

To make possible to correclty receive PDUs that optionally include
Constant Tone Extension after CRC the PHYEND event must be used
instead of END event. PHYEND event is generated by Radio peripheral
for actual end of a PDU. In case there is a CTE the PHYEND event
is generated at end of the CTE. If PDU does not have CTE, then
PHYEND event is generated at end of CRC. In this case it is the
same instant as END event.

Use of PHYEND event is required only when BLE 5.1 Receiving Constant
Tone Extensions feature is supported. Because it is the same event as
END for SOCs that supports Direction Finding Extension, it is always
used for that SOCs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: radio: add PHYEND delay compensation for nRF52

Radio peripheral that includes Direction Finding Extension is
able to generate EVENTS_PHYEND at very end of received packet.
In case there is a Constant Tone Extensions in received packet
the EVENTS_PHYEND event is generated after end of CTE reception.
If the CTE is not present in received PDU, the EVENTS_PHYEND
event is generate in the same instant as EVENTS_END.

If CTEINLINE is enabled then Radio will do a runtime packet parsing
to check if CTEInfo is present in received packet. In case there
is no CTEInfo the EVENTS_PHYEND event will be generated with
16 us delay after EVENTS_END.

To maintain IFS, additional EVENT_COMPARE is used that
will timeout earlier than regular EVENT_COMPARE for EVENTS_PHYEND
generated withtout delay. That additional EVENT_COMPARE will start
software switching of radio mode to TX. In case there is a CTEInfo
present in the packet, additional PPI wiring will cancell
EVENTS_COMPARE set for delayed EVENTS_PHYEND.

The commit add support for delayed PHYEND event for nRF52 SOCs.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: util: Add generic double buffer implementation

There are multiple places where double buffer is used in controlers
code. This commit adds generic implementation of the double buffer.
It can be used in future in all places where the data structure is
in use.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: LLL: allow setting CTEInfo to be in S1 byte

CTEInfo may be stored in S1 byte for BLE1M or BLE2M data channel
packet or in payload of advertising channel packet. To allow
selection of where to expect the CTEInfo to be Radio and
DF lower link layer API has to be changed. New parameter
cte_info_in_s1 was introduced.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: Add CTE recv and sample in connected mode

In connected mode when Receiving Constant Tone Extensions feature
is enabled, controller shall be able to receive CTE in any
data channel packet.
The commit adds required changes to allow receive of CTE for
all data channel packets in peripheral role.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: align DF conn less mode to use PHYEND event

Connectionless mode of direction finding was based on assumption
that PDU ends when EVENTS_END is generated.

Advertiser added CTE length to IFS duration to correctly maintain
IFS between consecutive PDUs.

Scanner was using separate API to configure radio to configure
software TX/RX mode switch to use PHYEND event instead of END.
Currently every SOC that has Direction Finding Extension in
Radio peripheral uses PHYEND event to mark end of PDU.
There is no need for separate API to configure software switch.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
tests: Bluetooth: df: Fix not building unit tests

There were changes done to controller code that cause
tests to do not build and fail during execution.
Changes are related with:
- modified code related with extended advertising ADI
  field handling
- added generic double buffer data structure that is
  used in controller
- moved code that was building only when DF is enabled
- added EVENTS_PHYEND that is not available in nrfbsim
  board

The PR addresses those issues.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: Controller: lll: Add IS_ENABLED to prevend build failures

Code that is part of radio_df_cte_inline_set_enabled may be not
compilable for targets that do not have Direction Finding Extension
in Radio peripheral. Added condition with IS_ENABLED to execute
the code only for those SOCs that have the extension.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Requires testing after rebase.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -470,6 +471,35 @@ uint8_t ll_df_set_cl_iq_sampling_enable(uint16_t handle,
+	df_cfg->last = 0U;
+}
+
+uint8_t ull_df_sync_cfg_is_disabled_or_requested_to_disable(struct lll_df_sync *df_cfg)
+{
+	struct lll_df_sync_cfg *cfg;
+
+	/* If new CTE sampling configuration was enqueued, get reference to
+	 * latest congiruation without swapping buffers. Buffer should be
+	 * swapped only at the beginning of the radio event.
+	 *
+	 * We may not get here if CTE sampling is not enabled in current
+	 * configuration.
+	 */
+	if (lll_df_sync_cfg_is_modified(df_cfg)) {
+		cfg = lll_df_sync_cfg_peek(df_cfg);
+	} else {
+		cfg = lll_df_sync_cfg_curr_get(df_cfg);
+	}
+
+	return !cfg->is_enabled;
+}
+#endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
```
Ah, that was for `ull_df_conn_cfg_is_not_enabled`. I feel like the same approach could be used here.

ppryga replied:
Yes, I didn't aligned it in this PR but I've forgotten that I moved this code.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df_types.h:
```diff
@@ -28,7 +28,15 @@ enum df_switch_sample_support {
+#define SYNC_IQ_REPORT_CNT (CONFIG_BT_PER_ADV_SYNC_MAX * CONFIG_BT_CTLR_DF_PER_SCAN_CTE_NUM_MAX * 2)
 #else
-#define IQ_REPORT_CNT 0
+#define SYNC_IQ_REPORT_CNT 0
```
Not sure what actually goes here, but should it have been `0U` instead? 

Ditto for `CONN_IQ_REPORT_CNT`

ppryga replied:
Its value is in range `0-16`, so yes it should be `0U`.

Thalley commented on subsys/bluetooth/controller/util/dbuf.h:
```diff
@@ -0,0 +1,107 @@
+ *
+ * Pointer to lates element is returned.
+ *
+ * @param hdr Pointer to double buffer header.
+ * @param[out] is_modifled Pointer to return information if buffer was swapped.
```
```suggestion
 * @param[in]  hdr Pointer to double buffer header.
 * @param[out] is_modifled Pointer to return information if buffer was swapped.
```

ppryga replied:
Isn't `[in]` the default?

Thalley replied:
`[in]` is usually omitted when all arguments are `in`, but when adding an `[out]` we usually add `[in]` for all `in` arguments as well. Doesn't seem to be a requirement, as we do also have `[out]`s without `[in]`, but it is more typical to add `[in]` as far as I can tell.

ppryga replied:
I'll add `[in]` and please lets stick to one convention.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.h:
```diff
@@ -5,6 +5,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+enum radio_end_evt_delay_state { END_EVT_DELAY_DISABLED, END_EVT_DELAY_ENABLED };
```
Yes it could be, but then it would have been better to use some macro instead of true/false. That would have made the code more readable in call locations e.g.:
`sw_switch(SW_SWITCH_RX, SW_SWITCH_TX, phy_rx, flags_rx, phy_tx, flags_tx, END_EVT_DELAY_DISABLED);`
instead of:
`sw_switch(SW_SWITCH_RX, SW_SWITCH_TX, phy_rx, flags_rx, phy_tx, flags_tx, false);`

Is it ok to leave the `enum`? Or do you still think that `boolean` is better here?

cvinayak replied:
I will stick to the enum, as the parameter value passed is readable.

Thalley replied:
I'm not going to block this PR on this, but arguing that named macros/enums should be used instead of boolean true/false is a bit odd to me. If it makes to do here, doesn't it make sense to do it everywhere, and avoid using the `bool` type anywhere?

Personally I don't see this specific function as being unique, so if you prefer named defines here for a boolean value (enable/disable, which we have a lot of places in the codebase), do you also prefer that for the other cases we have?

ppryga replied:
Well, for me it should be decided on particular case. E.g. in examples from my comment: https://github.com/zephyrproject-rtos/zephyr/pull/41406#discussion_r774108676 I prefer named constants instead of `true`, `false`. 
When I spot like like the one with `false` then I have to look into function declaration to check what is the argument purpose. `false` does not carry enough information for me. With named constant I have missing information.

Thalley replied:
In this case, we are talking about an enable/disable parameter, which we more often than not have used `bool` for in the code base. 

> When I spot line like the one with false then I have to look into function declaration to check what is the argument purpose. 

But isn't that true for every single parameter that isn't a named enum? For any `uintX_t` you'd also need to check the function declaration to know what it means.

ppryga replied:
> In this case, we are talking about an enable/disable parameter, which we more often than not have used bool for in the code base.

For me that depends on the function, it name and parameters list. e.g.  
```
void start_fsm(bool enable);

// somewhere else in a code 

start_fsm(true);
// or 
start_fsm(false);
```
is quite clear.

If a parameters list is longer then things become more unclear and one needs to look into declaration.

Thalley replied:
As mentioned, I will not block this PR, but I am still in favor using `bool` instead of a named enum for a boolean value.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -865,3 +949,47 @@ static void empty_tx_init(void)
+		iq_report->sample_count = radio_df_iq_samples_amount_get();
+		iq_report->packet_status = packet_status;
+		iq_report->rssi_ant_id = ant;
+		iq_report->cte_info = *(struct pdu_cte_info *)&cte_info;
```
Radio API uses mostly basic types. I didn't want to create a dependency on `pdu.h` in the API.

Thalley replied:
Can't argue with that :)

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -865,3 +949,47 @@ static void empty_tx_init(void)
 }
+
+#if defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+static inline bool create_iq_report(struct lll_conn *lll, uint8_t rssi_ready, uint8_t packet_status)
```
Changed

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -189,8 +201,15 @@ void lll_conn_isr_rx(void *param)
+#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_RX */
 	} else {
-		crc_ok = rssi_ready = 0U;
+		crc_ok = rssi_ready = 0;
```
Corrected

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2902,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+		sep->sample_count = 0;
+	} else {
+		for (idx = 0; idx < samples_cnt; ++idx) {
+			iq_tmp = IQ_SHIFT_12_TO_8_BIT(iq_report->sample[idx].i);
+			sep->sample[idx].i = (int8_t)iq_tmp;
+			iq_tmp = IQ_SHIFT_12_TO_8_BIT(iq_report->sample[idx].q);
+			sep->sample[idx].q = (int8_t)iq_tmp;
```
Changed `iq_tmp` type to `int8_t`. I've moved cast as suggested. There is narrowing conversion, so I'd like to be explicit here that it is desired.

Thalley replied:
> There is narrowing conversion, so I'd like to be explicit here that it is desired.

It could be argued that the name of `IQ_SHIFT_12_TO_8_BIT` is explicit enough (8 Bit is very accurate). Perhaps the macro itself needs to have the cast, rather than requiring the user of the macro to cast.

ppryga replied:
> It could be argued that the name of IQ_SHIFT_12_TO_8_BIT is explicit enough (8 Bit is very accurate)
Do you mean that the `IQ_SHIFT_12` part is not clear?

ppryga replied:
BTW I've moved the cast into macro.

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2902,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+			sep->sample[idx].i = (int8_t)iq_tmp;
+			iq_tmp = IQ_SHIFT_12_TO_8_BIT(iq_report->sample[idx].q);
+			sep->sample[idx].q = (int8_t)iq_tmp;
+		}
+		sep->sample_count = samples_cnt;
```
Empty line added.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral.c:
```diff
@@ -206,12 +213,53 @@ static int prepare_cb(struct lll_prepare_param *p)
+		} else {
+			lll_df_conf_cte_info_parsing_enable();
+		}
+#if defined(CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE)
+		end_evt_delay = END_EVT_DELAY_ENABLED;
+	} else {
+		end_evt_delay = END_EVT_DELAY_DISABLED;
+#endif /* CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE */
```
I was told by @cvinayak once to do not use `IS_ENABLED` in controller's code. I've been removing it from my commits in the past.

cvinayak replied:
Can not use `IS_ENABLED` here, `if (IS_ENABLED(...)) {...}` would fail compilation here as `end_evt_delay` is conditionally declared.

ppryga replied:
@cvinayak Thanks for checking the details here. I didn't look into that, because I thought that this is a kind of a policy to avoid IS_ENABLED in the Controllers code. Have I misunderstood something or has it changed?

cvinayak replied:
No, there is no policy to avoid IS_ENABLED. IS_ENABLED to be used where compilation is not an issue (like missing struct members or prior declarations). IS_ENABLED is not used in preprocessor conditional compilations (no `#if IS_ENABLED(...)` ) but `if (IS_ENABLED(...) {...}` in c control flow is preferred.

Thalley replied:
Agreed. `IS_ENABLED` should always be preferred if possible without adding any overhead. 

Vinayak is also correct in that `IS_ENABLED` cannot be used here though, as `end_evt_delay` is conditionally compiled (I had misread the code).

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral.c:
```diff
@@ -206,12 +213,53 @@ static int prepare_cb(struct lll_prepare_param *p)
+
+#if defined(CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE)
+/* Use special API for SOC that requires compensation for PHYEND event delay. */
 #if defined(CONFIG_BT_CTLR_PHY)
-	radio_switch_complete_and_tx(lll->phy_rx, 0, lll->phy_tx,
-				     lll->phy_flags);
+	radio_switch_complete_with_delay_compensation_and_tx(lll->phy_rx, 0, lll->phy_tx,
+							     lll->phy_flags, end_evt_delay);
 #else /* !CONFIG_BT_CTLR_PHY */
+	radio_switch_complete_with_delay_compensation_and_tx(0, 0, 0, 0, end_evt_delay);
+#endif /* !CONFIG_BT_CTLR_PHY */
+
+#endif /* CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE */
+#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_RX */
```
I was told by @cvinayak once to do not use IS_ENABLED in controller's code. I've been removing it from my commits in the past.

cvinayak replied:
Can not use `IS_ENABLED` here, `if (IS_ENABLED(...)) {...}` would fail compilation here as `end_evt_delay` is conditionally declared.

ppryga replied:
See: https://github.com/zephyrproject-rtos/zephyr/pull/41406/commits/1001697ecd9588d51a158aa641055db4bb9821f7#r774808840

Thalley replied:
Vinayak is here once again correct in that we cannot use IS_ENABLED. Several of the values are guarded in the struct.

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_peripheral.c:
```diff
@@ -206,12 +213,53 @@ static int prepare_cb(struct lll_prepare_param *p)
+ * - CTE RX is not enabled,
+ * - SOC does not require compensation for PHYEND event delay.
+ */
+#if !defined(CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE)
+#if defined(CONFIG_BT_CTLR_PHY)
+	radio_switch_complete_and_tx(lll->phy_rx, 0, lll->phy_tx, lll->phy_flags);
+#else /* !CONFIG_BT_CTLR_PHY && !CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE */
 	radio_switch_complete_and_tx(0, 0, 0, 0);
 #endif /* !CONFIG_BT_CTLR_PHY */
+#endif /* !CONFIG_BT_CTLR_DF_PHYEND_OFFSET_COMPENSATION_ENABLE */
```
I was told by @cvinayak once to do not use IS_ENABLED in controller's code. I've been removing it from my commits in the past.

cvinayak replied:
Here `if (IS_ENABLED(...)) {...}`  c construct can be used.

ppryga replied:
Added `IS_ENABLED` here. The code does not look amazing, at least for me, but it is checked during compilation.
Unfortunately parts guarded by `CONFIG_BT_CTLR_PHY` are not possible to be put into `IS_EANBLED` statement.

Thalley replied:
> Added `IS_ENABLED` here. The code does not look amazing, at least for me, but it is checked during compilation. Unfortunately parts guarded by `CONFIG_BT_CTLR_PHY` are not possible to be put into `IS_EANBLED` statement.

True, since e.g. `lll->phy_rx` is guarded by `CONFIG_BT_CTLR_PHY` in the struct definition.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1486,6 +1483,17 @@ void ll_rx_mem_release(void **node_rx)
 #endif /* CONFIG_BT_CTLR_SYNC_PERIODIC */
 
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX) || defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+		case NODE_RX_TYPE_SYNC_IQ_SAMPLE_REPORT:
```
Added `__fallthrough`

ppryga commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -1079,7 +1124,7 @@ uint8_t ll_df_set_conn_cte_req_enable(uint16_t handle, uint8_t enable, uint8_t c
 		 */
 	} else {
-		if (conn->df_rx_params.state == DF_CTE_SAMPLING_UNINITIALIZED) {
+		if (conn->df_rx_params.is_enabled == false) {
```
Changed

ppryga-nordic committed:
Bluetooth: controller: radio: add PHYEND delay compensation for nRF53

Radio peripheral that includes Direction Finding Extension is
able to generate EVENTS_PHYEND at very end of received PDU.
In case there is a Constant Tone Extensions in received packet
the EVENTS_PHYEND event is generated after end of CTE reception.
If the CTE is present in received packet, the EVENTS_PHYEND
event is generate in the same instant as EVENTS_END.

If CTEINLINE is enabled then Radio will do a runtime packet parsing
to check if CTEInfo is present in received packet. In case there
is no CTEInfo the EVENTS_PHYEND event will be generated with
16 us delay after EVENTS_END.

To maintain the IFS, additional EVENT_COMPARE is used that
will timeout earlier than regular EVENT_COMPARE for EVENTS_PHYEND
generated withtou delay. That additional EVENT_COMPARE will start
software switching of radio mode to TX. In case there is a CTEInfo
present in the packet, additional PPI wiring will cancell
EVENTS_COMPARE set for delayed EVENTS_PHYEND.

The commit provides changes to support delayed PHYEND for nRF53 SOCs.

Besides that there are small formatting corrections due to extended
max line length.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio.c.orig:
```diff
@@ -0,0 +1,1679 @@
+ * Copyright (c) 2016 Vinayak Kariappa Chettimada
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Yes, it was a mistake

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_nrf5_resources.h.orig:
```diff
@@ -0,0 +1,119 @@
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
```
Nope it was a mistake

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync.c:
```diff
@@ -283,16 +280,18 @@ static int create_prepare_cb(struct lll_prepare_param *p)
 #endif /* CONFIG_BT_CTLR_DF_SCAN_CTE_RX */
+/* Guarded separately to allow to compile-in the code when direction finding is disabled */
+#if defined(CONFIG_BT_CTLR_DF_SUPPORT)
 	{
 		radio_df_cte_inline_set_enabled(false);
-		radio_switch_complete_and_disable();
 	}
+#endif /* CONFIG_BT_CTLR_DF_SUPPORT */
```
Same as above

Thalley replied:
This should still be able to use IS_ENABLED.

ppryga replied:
Nope, the `radio_df_cte_inline_set_enabled` is compiled conditionally.

Thalley replied:
But isn't it included from a header? 
If so, then IS_ENABLED can still be used, unless you have guarded the definition in the header file as well.

ppryga replied:
You are right. I've reworked the code around the line to use IS_ENABLED and make it similar to other places where `if (0) {` is used in LL.

ppryga commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2431,17 +2439,20 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+#endif /* CONFIG_BT_OBSERVER */
+
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX) || defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+	case NODE_RX_TYPE_SYNC_IQ_SAMPLE_REPORT:
```
Added `__fallthrough;`

cvinayak replied:
As there is no specific code fallthrough, it is not necessary. I would exceptionally add (when there is no code) when checkpatch script misundertands due to compilation compilations of the `case`.

ppryga replied:
I haven't read about fallthrough before, hence I assumed it as a good practice e.g. in host. 
After some reading, I agree with @cvinayak here is is not necessary. There is no code executed for `case NODE_RX_TYPE_SYNC_IQ_SAMPLE_REPORT`, hence it is not necessary to use the statement.

BTW. Since C23 release the `[[ fallthrough ]]` or `[[ __fallthrough__ ]]` attribute will be part of standard.

ppryga replied:
Removed `__fallthrough` from places where it was added.

Thalley replied:
Looks like @cvinayak is correct that if there is no code before the fallthrough, this can be omitted. From searching in the code base, it does look like `__fallthrough` is still used many places without any code, but it does not seem to be something required by the Linux kernel style (which we also follow).

ppryga commented:
Rebased and addressed comments.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -88,7 +94,7 @@ struct node_rx_iq_report {
 struct lll_df_sync_cfg {
 	uint8_t is_enabled:1;
-	uint8_t slot_durations:2; /* One of possible values: 1us, 2us. */
+	uint8_t slot_durations:2; /* Bit field where: BIT(1) is 1us, BIT(2) is 2us. */
```
```suggestion
	uint8_t slot_durations:2; /* Bit field where: BIT(0) is 1us, BIT(1) is 2us. */
```

Are there 3 values hence need 2 bits? or is it bitmask where both 1us and 2us is allowed? If only one of 1us or 2us is allow why not use 1-bit field?

ppryga replied:
BT 5.3 Core spec defines slot durations as: `0x01 Switching and sampling slots are 1 μs each` and `0x02 Switching and sampling slots are 2 μs each`. To avoid translation of the values (and extra operations) I decided to use one more bit and stick to definitions and assigned values. This also avoids confusion when reading the code executed from host down to LLL. 

BTW. Even if I change it to use single bit, the size of the struct will be the same.

Thalley replied:
Vinayak is still correct that if you use `BIT(1)` and `BIT(2)`, then we need `uint8_t slot_durations:2` as `BIT(2)` would set the 3rd bit (value 4).

ppryga replied:
Ah, I've omitted it, thanks!

cvinayak commented on subsys/bluetooth/controller/ll_sw/pdu.h:
```diff
@@ -830,6 +830,7 @@ struct pdu_data {
 		uint8_t resv; /* TODO: remove nRF specific code */
 		struct pdu_cte_info cte_info; /* BT 5.1 Core spec. CTEInfo storage */
 	};
+
```
redundant blank line

ppryga replied:
Removed

ppryga commented:
Added changes after review by @cvinayak.
Rebased on main to resolve conflicts.

ppryga commented:
@cvinayak and @Thalley it looks that all comments were addressed. Please take a look.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -1486,6 +1483,17 @@ void ll_rx_mem_release(void **node_rx)
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX) || defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+		case NODE_RX_TYPE_SYNC_IQ_SAMPLE_REPORT:
+		case NODE_RX_TYPE_CONN_IQ_SAMPLE_REPORT:
+		{
+			ull_iq_report_link_inc_quota(1);
+			ull_df_iq_report_mem_release(rx_free);
+			ull_df_rx_iq_report_alloc(1);
+		}
```
Either
```suggestion
			ull_iq_report_link_inc_quota(1);
			ull_df_iq_report_mem_release(rx_free);
			ull_df_rx_iq_report_alloc(1);
```
or
```suggestion
		{
			const uint8_t report_cnt = 1U;
			
			ull_iq_report_link_inc_quota(report_cnt);
			ull_df_iq_report_mem_release(rx_free);
			ull_df_rx_iq_report_alloc(report_cnt);
		}
```

ppryga replied:
Neither of these. I leave it as is.
Braces are left to follow convention in this function.
Const variable is not necessary here. In my humble opinion `ull_iq_report_link_inc_quota` and `ull_df_rx_iq_report_alloc` with single argument being integer is enough to understand meaning of provided values.

Thalley replied:
Ok :) My rule of thumb is that if a literal value is used more than once for the same value (e.g. number of reports), then always put in a variable. Not only does it make it easier to understand that the value `1` in both function calls is the same value, but it also makes it easier to refactor.

Given that it is a const, it will also likely be optimized by the compiler to be the same result, so I usually see such constant values as a win-win :) 

It is OK to keep as is, although I am curious to when you would use a constant instead of literals: 3 or more function calls, or never?

ppryga replied:
Your argument about multiple use of a value with the same meaning convinced me. I've applied suggested change.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -2431,17 +2439,20 @@ static inline int rx_demux_rx(memq_link_t *link, struct node_rx_hdr *rx)
+#if defined(CONFIG_BT_CTLR_DF_SCAN_CTE_RX) || defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+	case NODE_RX_TYPE_SYNC_IQ_SAMPLE_REPORT:
+	case NODE_RX_TYPE_CONN_IQ_SAMPLE_REPORT:
+	{
```
I think you can also omit the `{` for this case, right?

ppryga replied:
It could be committed but to follow the convention in this function I leave it as is.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_df_internal.h:
```diff
@@ -20,4 +20,6 @@ void ull_df_rx_iq_report_alloc(uint8_t max);
 void ull_df_sync_cfg_init(struct lll_df_sync *cfg);
 /* Returns information if CTE sampling for periodic sync is requested to disable. */
-uint8_t ull_df_sync_cfg_is_disabled_or_requested_to_disable(struct lll_df_sync *df_cfg);
+bool ull_df_sync_cfg_is_not_enabled(struct lll_df_sync *df_cfg);
+/* Returns information if CTE sampling for a connection is not enabled. */
+bool ull_df_conn_cfg_is_not_enabled(struct lll_df_conn_rx_cfg *rx_cfg);
```
Given that these are `get`ters, the arguments could and should be made `const`.

ppryga replied:
These can't be changed to `const` because the functions use e.g. 
```
static inline void *dbuf_peek(struct dbuf_hdr *hdr)
{
	return &hdr->data[hdr->last * hdr->elem_size];
}
```
that can't have `const struct dbuf_hdr *hdr` due to return `void*`. That would limit the usage of the `dbuf_peek` funcion.

Thalley replied:
I see :)

Thalley commented on subsys/bluetooth/controller/util/dbuf.h:
```diff
@@ -0,0 +1,107 @@
+ *
+ * @return True if there were new element enqueued, false in other case.
+ */
+static inline bool dbuf_is_modified(struct dbuf_hdr *hdr)
```
```suggestion
static inline bool dbuf_is_modified(const struct dbuf_hdr *hdr)
```

ppryga replied:
`const` added

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_internal.h:
```diff
@@ -57,7 +57,10 @@ static inline uint8_t lll_df_sync_cfg_is_modified(struct lll_df_sync *df_cfg)
 /* Enables CTE reception according to provided configuration */
 void lll_df_conf_cte_rx_enable(uint8_t slot_duration, uint8_t ant_num, uint8_t *ant_ids,
-			       uint8_t chan_idx);
+			       uint8_t chan_idx, bool cte_info_in_s1);
```
`const` added

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+	if (iq_report->packet_status == BT_HCI_LE_CTE_INSUFFICIENT_RESOURCES) {
+		sep->sample[0].i = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
+		sep->sample[0].q = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
+		sep->sample_count = 0;
```
Corrected

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+		sep->sample[0].q = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
+		sep->sample_count = 0;
+	} else {
+		for (idx = 0; idx < samples_cnt; ++idx) {
```
Corrected

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+		sep->sample[0].q = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
+		sep->sample_count = 0;
+	} else {
+		for (idx = 0; idx < samples_cnt; ++idx) {
```
Changed to limit the scope of variable

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -435,6 +452,10 @@ void lll_conn_isr_rx(void *param)
+#if defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+	struct lll_df_conn_rx_params *df_rx_params;
+	struct lll_df_conn_rx_cfg *df_rx_cfg;
+#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_RX */
```
Variables moved to limit the scope

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -450,13 +471,55 @@ void lll_conn_isr_tx(void *param)
+		 */
+		df_rx_params = dbuf_curr_get(&df_rx_cfg->hdr);
+
+		if (df_rx_params->is_enabled == true) {
```
comparison removed

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -473,9 +536,10 @@ void lll_conn_isr_tx(void *param)
 	/* +/- 2us active clock jitter, +1 us hcto compensation */
 	hcto = radio_tmr_tifs_base_get() + EVENT_IFS_US + (EVENT_CLOCK_JITTER_US << 1) +
-	       RANGE_DELAY_US + HCTO_START_DELAY_US;
+	       RANGE_DELAY_US + 1;
```
Good catch, I've didn't spot that while resolving conflicts after rebase. Thank you!

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -929,6 +998,50 @@ static void empty_tx_init(void)
+	struct lll_df_conn_rx_cfg *rx_cfg;
+	struct node_rx_ftr *ftr;
+	uint8_t cte_info;
+	uint8_t ant;
```
Variables moved to limit the scope

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -60,7 +66,7 @@ struct lll_df_adv_cfg {
 
 #define RSSI_DBM_TO_DECI_DBM(x) (-(x) * 10)
-#define IQ_SHIFT_12_TO_8_BIT(x) ((x) >> 4)
+#define IQ_SHIFT_12_TO_8_BIT(x) (int8_t)((x) >> 4)
```
Brackets added

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -102,24 +108,33 @@ struct lll_df_sync {
+	/* Channel is set only once for a connection vent. The information will be used during CTE
+	 * RX configuration by ISR handlers.
+	 */
+	uint8_t chan;
```
`chan` is a channel number so it should be in a range 0-40. 6 bits are required to store this range. I've changed that.

ppryga commented:
@Thalley changes implemented, thank you for another good review.

ppryga commented:
Rebased on main branch. There were some issue in CI related with posting results.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -402,7 +406,20 @@ void lll_conn_isr_rx(void *param)
+		is_iq_report = false;
+	}
+
+	if (is_ull_rx || is_iq_report == true) {
```
```suggestion
	if (is_ull_rx || is_iq_report) {
```

ppryga replied:
Corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -491,11 +554,16 @@ void lll_conn_isr_tx(void *param)
 	radio_tmr_hcto_configure(hcto);
 
-#if defined(CONFIG_BT_CENTRAL) && defined(CONFIG_BT_CTLR_CONN_RSSI)
+#if defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
+	if (!trx_cnt) {
+#elif defined(CONFIG_BT_CENTRAL) && defined(CONFIG_BT_CTLR_CONN_RSSI)
 	if (!trx_cnt && !lll->role) {
+#else
+	if (0) {
+#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_RX */
+
```
Dont we want rssi for all Rx-ed PDU?

```suggestion
#if defined(CONFIG_BT_CTRL_DF_CONN_CTE_RX)
	if (true) {
#elif defined(CONFIG_BT_CENTRAL) && defined(CONFIG_BT_CTLR_CONN_RSSI)
	if (!trx_cnt && !lll->role) {
#else
	if (0) {
#endif /* CONFIG_BT_CTRL_DF_CONN_CTE_RX */

```

Thalley replied:
Isn't it also a bit odd to use `0` for false, and `true` for true, instead of using true/false, or 1/0?

ppryga replied:
Corrected with use of true/false in both if statements.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -929,6 +997,51 @@ static void empty_tx_init(void)
+
+	rx_params = dbuf_curr_get(&rx_cfg->hdr);
+
+	if (rx_params->is_enabled == true) {
```
```suggestion
	if (rx_params->is_enabled) {
```

ppryga replied:
Corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df_types.h:
```diff
@@ -102,24 +108,33 @@ struct lll_df_sync {
+	 * It is required for handling HCI_LE_Connection_CTE_Request_Enable HCI command.
+	 * See BT 5.3 Core specification Vol 4, Part E, sec. 7.8.85.
+	 */
+	bool is_initialized:1;
```
Any reason for `uint8_t is_enabled:1;` but here `bool is_initialized:1;` ? Use `uint8_t is_initialized:1;`

Thalley replied:
If we want to use the same 8-bits for both values, don't we need to do that as well?
i.e.
```
uint8_t is_initialized:1;
uint8_t chan:6;
```

Or can you mix and match `bool` and `uint8_t`?

ppryga replied:
I think that I was asked about such change before. I'm not sure if I were requested to use `bool` in such situation.
Changed to `uint8_t`.

Thalley replied:
I think I requested it to be a bool, but now that `chan` is 6 bits, both of them can fit in a single 8-bit value.

ppryga replied:
> Or can you mix and match bool and uint8_t?

I didn't find clear information to answer the question. I've checked it with gcc 11 and the output was as expected, mix of types didn't increase the size of the structure. 

Ref: https://en.cppreference.com/w/c/language/bit_field

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2847,12 +2847,15 @@ static void le_df_connectionless_iq_report(struct pdu_data *pdu_rx,
 		sep->sample[0].q = BT_HCI_LE_CTE_REPORT_NO_VALID_SAMPLE;
 		sep->sample_count = 0;
 	} else {
+		int8_t iq_tmp;
+
```
We could also possibly avoid the tmp variable totally, as 
```
			sep->sample[idx].i = IQ_SHIFT_12_TO_8_BIT(iq_report->sample[idx].i);
			sep->sample[idx].q = IQ_SHIFT_12_TO_8_BIT(iq_report->sample[idx].q);
```
isn't too bad to read

ppryga replied:
The variable is removed completely. It was used to avoid line breaking due to 80 characters max line length.

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+{
+	struct bt_hci_evt_le_connection_iq_report *sep;
+	struct node_rx_iq_report *iq_report;
+
```
Removed

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+
+	phy_rx = lll->phy_rx;
+	/* Make sure the report is generated for connection on PHY UNCODED */
+	LL_ASSERT(lll->phy_rx != PHY_CODED);
```
Corrected

ppryga commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -2901,6 +2904,86 @@ static void le_df_set_conn_cte_rx_params(struct net_buf *buf, struct net_buf **e
+
+	struct lll_conn *lll;
+	uint8_t samples_cnt;
+	int16_t iq_tmp;
```
The variable is removed completely. It was used to avoid line breaking due to 80 characters max line length.

ppryga commented:
@Thalley @cvinayak changes done.

Thalley reviewed (approved):
All my comments have been addressed.
